File: packages/motia/src/guards.ts
import type { Step, TriggerConfig } from './types'

// Type aliases to avoid deep type instantiation
type ApiTriggerType = Extract<TriggerConfig, { type: 'api' }>
type EventTriggerType = Extract<TriggerConfig, { type: 'event' }>
type CronTriggerType = Extract<TriggerConfig, { type: 'cron' }>
type StateTriggerType = Extract<TriggerConfig, { type: 'state' }>
type StreamTriggerType = Extract<TriggerConfig, { type: 'stream' }>

export const isApiTrigger = (trigger: TriggerConfig): trigger is ApiTriggerType => trigger.type === 'api'

export const isEventTrigger = (trigger: TriggerConfig): trigger is EventTriggerType => trigger.type === 'event'

export const isCronTrigger = (trigger: TriggerConfig): trigger is CronTriggerType => trigger.type === 'cron'

export const isStateTrigger = (trigger: TriggerConfig): trigger is StateTriggerType => trigger.type === 'state'

export const isStreamTrigger = (trigger: TriggerConfig): trigger is StreamTriggerType => trigger.type === 'stream'

export const getApiTriggers = (step: Step): ApiTriggerType[] => step.config.triggers.filter(isApiTrigger)

export const getEventTriggers = (step: Step): EventTriggerType[] => step.config.triggers.filter(isEventTrigger)

export const getCronTriggers = (step: Step): CronTriggerType[] => step.config.triggers.filter(isCronTrigger)

File: motia-py/packages/motia/src/motia/guards.py
"""Type guards for step and trigger configurations."""

from .types import ApiTrigger, CronTrigger, QueueTrigger, StateTrigger, Step, StreamTrigger, TriggerConfig


# Trigger-level guards (matching Node.js)
def is_api_trigger(trigger: TriggerConfig) -> bool:
    """Check if a trigger is an API trigger."""
    return isinstance(trigger, ApiTrigger)


def is_queue_trigger(trigger: TriggerConfig) -> bool:
    """Check if a trigger is a queue trigger."""
    return isinstance(trigger, QueueTrigger)


def is_cron_trigger(trigger: TriggerConfig) -> bool:
    """Check if a trigger is a cron trigger."""
    return isinstance(trigger, CronTrigger)


def is_state_trigger(trigger: TriggerConfig) -> bool:
    """Check if a trigger is a state trigger."""
    return isinstance(trigger, StateTrigger)


def is_stream_trigger(trigger: TriggerConfig) -> bool:
    """Check if a trigger is a stream trigger."""
    return isinstance(trigger, StreamTrigger)


# Step-level guards
def is_api_step(step: Step) -> bool:
    """Check if a step has an API trigger."""
    return any(is_api_trigger(t) for t in step.config.triggers)


def is_queue_step(step: Step) -> bool:
    """Check if a step has a queue trigger."""
    return any(is_queue_trigger(t) for t in step.config.triggers)


def is_cron_step(step: Step) -> bool:
    """Check if a step has a cron trigger."""
    return any(is_cron_trigger(t) for t in step.config.triggers)


# Getter helpers (matching Node.js)
def get_api_triggers(step: Step) -> list[ApiTrigger]:
    """Get all API triggers from a step."""
    return [t for t in step.config.triggers if isinstance(t, ApiTrigger)]


def get_queue_triggers(step: Step) -> list[QueueTrigger]:
    """Get all queue triggers from a step."""
    return [t for t in step.config.triggers if isinstance(t, QueueTrigger)]


def get_cron_triggers(step: Step) -> list[CronTrigger]:
    """Get all cron triggers from a step."""
    return [t for t in step.config.triggers if isinstance(t, CronTrigger)]
